test/CodeGenOpenCL/sampler_t.cl

```diff
@@ -0,0 +1,68 @@
+// RUN: %clang_cc1 %s -triple spir-unknown-unknown -emit-llvm -o - -O0 | FileCheck %s
+
+// CHECK: %opencl.sampler_t = type { i32 }
+
+#define CLK_NORMALIZED_COORDS_TRUE 0x01
+#define CLK_ADDRESS_REPEAT 0x02
+#define CLK_FILTER_NEAREST 0x04
+
+const sampler_t gs = 2;
+// CHECK: @gs = constant %opencl.sampler_t { i32 2 }, align 4
+
+constant sampler_t cgs = 3;
+// CHECK: @cgs = addrspace(2) constant %opencl.sampler_t { i32 3 }, align 4
+
+// CHECK: @bar.kst = internal constant %opencl.sampler_t { i32 5 }, align 4
+// CHECK: @bar.cst = internal addrspace(2) constant %opencl.sampler_t { i32 6 }, align 4
+// CHECK: @bar.lst = internal constant %opencl.sampler_t { i32 7 }, align 4
+// CHECK: @bar.dd = internal addrspace(2) constant %opencl.sampler_t { i32 3 }, align 4
+
+
+void foo(const sampler_t st){}
+// CHECK: define spir_func void @foo(%opencl.sampler_t* byval %st)
+
+kernel void bar(const sampler_t ast) {
+// CHECK: define spir_kernel void @bar(%opencl.sampler_t* byval %ast)
+
+// CHECK: {{.*}} = alloca %opencl.sampler_t, align 4
+// CHECK: [[Q_SAMPLER:%[A-Za-z0-9_\.]+]] = alloca %opencl.sampler_t, align 4
+// CHECK: [[LITERAL_ARG:%[A-Za-z0-9_\.]+]] = alloca %opencl.sampler_t, align 4
+// CHECK: [[DD_SAMPLER:%[A-Za-z0-9_\.]+]] = alloca %opencl.sampler_t, align 4
+// CHECK: [[LITERAL_ARG2:%[A-Za-z0-9_\.]+]] = alloca %opencl.sampler_t, align 4
+
+    const sampler_t kst = 5;
+    foo(kst);
+// CHECK: call spir_func void @foo(%opencl.sampler_t* byval @bar.kst)
+
+    constant sampler_t cst = 6;
+    const sampler_t ost = cst;
+// CHECK: [[OST:%[A-Za-z0-9_\.]+]] = bitcast %opencl.sampler_t* %ost to i8*
+// CHECK: call void @llvm.memcpy.p0i8.p2i8.i32(i8* [[OST]], i8 addrspace(2)* bitcast (%opencl.sampler_t addrspace(2)* @bar.cst to i8 addrspace(2)*), i32 4, i32 4, i1 false)
```

Comment:
we should disable translation of load/store to memcpy for sampler since this introduces assumptions of the size of sampler which is platform dependent. Also the representation of sampler should allow compile-time determination of all constant sampler values at their use, i.e., each sampler argument passed to image read functions should be able to be traced to either a sampler kernel argument, or a literal value. memcpy of sampler makes this more difficult.